File: backend/routes/logo.js
const express = require('express');
const router = express.Router();
const multer = require('multer');
const path = require('path');
const fs = require('fs');
const crypto = require('crypto');
const uuidv4 = () => crypto.randomUUID();
const { auth } = require('../middleware/auth');
const { logoGenerationLimiter, incrementGenerationCount, getGenerationStatus } = require('../middleware/logoLimiter');
const { generateLogo } = require('../utils/logoGenerator');
const { deleteLogo } = require('../utils/logoCleanup');
const User = require('../models/User');

// Ensure logos directory exists
const logosDir = path.join(__dirname, '..', 'uploads', 'logos');
if (!fs.existsSync(logosDir)) {
  fs.mkdirSync(logosDir, { recursive: true });
}

// Configure Multer for logo uploads
const storage = multer.diskStorage({
  destination: (req, file, cb) => {
    cb(null, logosDir);
  },
  filename: (req, file, cb) => {
    const uniqueSuffix = `${req.user._id}-${Date.now()}-${uuidv4()}`;
    const ext = path.extname(file.originalname).toLowerCase();
    cb(null, `${uniqueSuffix}-custom${ext}`);
  }
});

const fileFilter = (req, file, cb) => {
  const allowedTypes = ['image/jpeg', 'image/png', 'image/svg+xml'];
  if (allowedTypes.includes(file.mimetype)) {
    cb(null, true);
  } else {
    cb(new Error('Only PNG, JPG, and SVG files are allowed'), false);
  }
};

const upload = multer({
  storage,
  fileFilter,
  limits: {
    fileSize: 2 * 1024 * 1024 // 2MB limit
  }
});

/**
 * POST /api/logo/generate
 * Generate a new logo using AI
 */
router.post('/generate', auth, logoGenerationLimiter, async (req, res) => {
  try {
    const { prompt } = req.body;

    // Validate prompt
    if (!prompt || typeof prompt !== 'string' || prompt.trim().length === 0) {
      return res.status(400).json({
        success: false,
        error: 'Invalid prompt',
        message: 'Please provide a valid logo description'
      });
    }

    if (prompt.length > 500) {
      return res.status(400).json({
        success: false,
        error: 'Prompt too long',
        message: 'Prompt must be less than 500 characters'
      });
    }

    // Generate unique filename
    const logoId = uuidv4();
    const baseFilename = `${req.user._id}-${Date.now()}-${logoId}`;
    const pngFilename = `${baseFilename}.png`;
    const outputPath = path.join(logosDir, pngFilename);

    // Generate logo using AI
    const generatedPath = await generateLogo(prompt.trim(), outputPath);

    // Determine actual filename and extension
    const actualExt = path.extname(generatedPath);
    const actualFilename = `${baseFilename}${actualExt}`;
    const relativePath = path.join('uploads', 'logos', actualFilename);
    const urlPath = `/uploads/logos/${actualFilename}`;

    // Increment user's generation count
    await incrementGenerationCount(req.user._id);

    // Save to user's generated logos
    const user = await User.findById(req.user._id);
    const logoEntry = {
      logoId: logoId, // Custom ID for reference
      url: urlPath,
      prompt: prompt.trim(),
      filePath: relativePath,
      createdAt: new Date(),
      expiresAt: new Date(Date.now() + 7 * 24 * 60 * 60 * 1000) // 7 days
    };

    user.generatedLogos.push(logoEntry);

    // Keep only last 20 logos in history
    if (user.generatedLogos.length > 20) {
      const removed = user.generatedLogos.shift();
      // Delete old file
      try {
        const oldPath = path.join(process.cwd(), removed.filePath);
        if (fs.existsSync(oldPath)) {
          fs.unlinkSync(oldPath);
        }
      } catch (err) {
        console.error('Error deleting old logo file:', err);
      }
    }

    await user.save();

    // Get updated generation status
    const status = await getGenerationStatus(req.user._id);

    res.json({
      success: true,
      logo: logoEntry,
      remainingGenerations: status.remaining,
      message: 'Logo generated successfully'
    });

  } catch (error) {
    console.error('Logo generation error:', error);
    console.error('Stack trace:', error.stack);
    res.status(500).json({
      success: false,
      error: 'Generation failed',
      message: error.message || 'Failed to generate logo. Please try again.',
      details: process.env.NODE_ENV === 'development' ? error.stack : undefined
    });
  }
});

/**
 * GET /api/logo/history
 * Get user's logo generation history
 */
router.get('/history', auth, async (req, res) => {
  try {
    const user = await User.findById(req.user._id)
      .select('generatedLogos businessLogo hasCustomLogo');

    if (!user) {
      return res.status(404).json({ message: 'User not found' });
    }

    // Filter out expired logos
    const now = new Date();
    const validLogos = user.generatedLogos.filter(
      logo => new Date(logo.expiresAt) > now
    );

    res.json({
      success: true,
      logos: validLogos.reverse(), // Most recent first
      businessLogo: user.businessLogo,
      hasCustomLogo: user.hasCustomLogo,
      totalGenerated: user.generatedLogos.length
    });

  } catch (error) {
    console.error('Error fetching logo history:', error);
    res.status(500).json({
      success: false,
      error: 'Failed to fetch logo history'
    });
  }
});

/**
 * GET /api/logo/status
 * Get user's generation limit status
 */
router.get('/status', auth, async (req, res) => {
  try {
    const status = await getGenerationStatus(req.user._id);

    if (!status) {
      return res.status(404).json({ message: 'User not found' });
    }

    res.json({
      success: true,
      status
    });

  } catch (error) {
    console.error('Error fetching generation status:', error);
    res.status(500).json({
      success: false,
      error: 'Failed to fetch generation status'
    });
  }
});

/**
 * POST /api/logo/reset-limit
 * Reset daily generation limit (for testing only)
 */
router.post('/reset-limit', auth, async (req, res) => {
  try {
    const user = await User.findById(req.user._id);
    if (!user) {
      return res.status(404).json({ message: 'User not found' });
    }

    user.logoGenerationCount.count = 0;
    user.logoGenerationCount.lastResetDate = new Date();
    await user.save();

    const { DAILY_LIMIT } = require('../middleware/logoLimiter');

    res.json({
      success: true,
      message: 'Logo generation limit reset successfully',
      status: {
        limit: DAILY_LIMIT,
        used: 0,
        remaining: DAILY_LIMIT,
        resetTime: (() => {
          const t = new Date();
          t.setUTCHours(24, 0, 0, 0);
          return t.toISOString();
        })(),
      }
    });
  } catch (error) {
    console.error('Error resetting limit:', error);
    res.status(500).json({ success: false, error: 'Failed to reset limit' });
  }
});

/**
 * PUT /api/logo/select/:logoId
 * Select a logo as the business logo
 */
router.put('/select/:logoId', auth, async (req, res) => {
  try {
    const { logoId } = req.params;
    const user = await User.findById(req.user._id);

    if (!user) {
      return res.status(404).json({ message: 'User not found' });
    }

    // Find the logo in generated logos (check both logoId and _id for backward compat)
    const logo = user.generatedLogos.find(l => l.logoId === logoId || l._id.toString() === logoId);

    if (!logo) {
      return res.status(404).json({
        success: false,
        error: 'Logo not found',
        message: 'The requested logo was not found in your history'
      });
    }

    // Check if logo has expired
    if (new Date(logo.expiresAt) < new Date()) {
      return res.status(400).json({
        success: false,
        error: 'Logo expired',
        message: 'This logo has expired. Please generate a new one.'
      });
    }

    // Set as business logo
    user.businessLogo = logo.url;
    await user.save();

    res.json({
      success: true,
      message: 'Logo selected as business logo',
      businessLogo: user.businessLogo
    });

  } catch (error) {
    console.error('Error selecting logo:', error);
    res.status(500).json({
      success: false,
      error: 'Failed to select logo'
    });
  }
});

/**
 * DELETE /api/logo/:logoId
 * Delete a specific logo
 */
router.delete('/:logoId', auth, async (req, res) => {
  try {
    const { logoId } = req.params;

    await deleteLogo(req.user._id, logoId);

    res.json({
      success: true,
      message: 'Logo deleted successfully'
    });

  } catch (error) {
    console.error('Error deleting logo:', error);
    res.status(500).json({
      success: false,
      error: 'Failed to delete logo',
      message: error.message
    });
  }
});

/**
 * POST /api/logo/upload
 * Upload a custom logo
 */
router.post('/upload', auth, upload.single('logo'), async (req, res) => {
  try {
    if (!req.file) {
      return res.status(400).json({
        success: false,
        error: 'No file uploaded',
        message: 'Please select a file to upload'
      });
    }

    const user = await User.findById(req.user._id);
    if (!user) {
      // Delete uploaded file if user not found
      fs.unlinkSync(req.file.path);
      return res.status(404).json({ message: 'User not found' });
    }

    // Delete previous custom logo if exists
    if (user.businessLogo && user.hasCustomLogo) {
      try {
        const oldPath = path.join(process.cwd(), 'uploads', user.businessLogo.replace('/uploads/', ''));
        if (fs.existsSync(oldPath)) {
          fs.unlinkSync(oldPath);
        }
      } catch (err) {
        console.error('Error deleting old custom logo:', err);
      }
    }

    // Set as business logo
    const relativePath = path.join('uploads', 'logos', req.file.filename);
    user.businessLogo = `/uploads/logos/${req.file.filename}`;
    user.hasCustomLogo = true;
    await user.save();

    res.json({
      success: true,
      message: 'Logo uploaded successfully',
      businessLogo: user.businessLogo
    });

  } catch (error) {
    console.error('Logo upload error:', error);

    // Delete uploaded file if error
    if (req.file && fs.existsSync(req.file.path)) {
      fs.unlinkSync(req.file.path);
    }

    res.status(500).json({
      success: false,
      error: 'Upload failed',
      message: error.message || 'Failed to upload logo'
    });
  }
});

/**
 * DELETE /api/logo/custom
 * Remove custom logo and revert to AI-generated or none
 */
router.delete('/custom/remove', auth, async (req, res) => {
  try {
    const user = await User.findById(req.user._id);
    if (!user) {
      return res.status(404).json({ message: 'User not found' });
    }

    if (user.hasCustomLogo && user.businessLogo) {
      // Delete the custom logo file
      try {
        const filePath = path.join(process.cwd(), 'uploads', user.businessLogo.replace('/uploads/', ''));
        if (fs.existsSync(filePath)) {
          fs.unlinkSync(filePath);
        }
      } catch (err) {
        console.error('Error deleting custom logo:', err);
      }
    }

    // Check if user has any AI-generated logos to fall back to
    const now = new Date();
    const validLogos = user.generatedLogos.filter(l => new Date(l.expiresAt) > now);

    if (validLogos.length > 0) {
      // Use most recent AI-generated logo
      user.businessLogo = validLogos[validLogos.length - 1].url;
    } else {
      user.businessLogo = null;
    }

    user.hasCustomLogo = false;
    await user.save();

    res.json({
      success: true,
      message: 'Custom logo removed',
      businessLogo: user.businessLogo
    });

  } catch (error) {
    console.error('Error removing custom logo:', error);
    res.status(500).json({
      success: false,
      error: 'Failed to remove custom logo'
    });
  }
});

// Error handling middleware for multer
router.use((error, req, res, next) => {
  if (error instanceof multer.MulterError) {
    if (error.code === 'LIMIT_FILE_SIZE') {
      return res.status(400).json({
        success: false,
        error: 'File too large',
        message: 'Logo must be less than 2MB'
      });
    }
    return res.status(400).json({
      success: false,
      error: 'Upload error',
      message: error.message
    });
  }

  if (error) {
    return res.status(400).json({
      success: false,
      error: 'Invalid file',
      message: error.message
    });
  }

  next();
});

module.exports = router;

File: go-backend/internal/handlers/logo.go
package handlers

import (
	"context"
	"crypto/rand"
	"encoding/hex"
	"fmt"
	"io"
	"log"
	"net/http"
	"net/url"
	"os"
	"path/filepath"
	"strings"
	"time"

	"github.com/gin-gonic/gin"
	"go.mongodb.org/mongo-driver/bson"
	"go.mongodb.org/mongo-driver/bson/primitive"
	"msme-marketplace/internal/database"
	"msme-marketplace/internal/models"
)

type LogoHandler struct {
	DailyLimit int
}

func NewLogoHandler() *LogoHandler {
	return &LogoHandler{DailyLimit: 5}
}

func (h *LogoHandler) GenerateLogo(c *gin.Context) {
	userID := c.GetString("userID")
	userObjID, err := primitive.ObjectIDFromHex(userID)
	if err != nil {
		c.JSON(400, gin.H{"message": "Invalid user ID"})
		return
	}

	var req struct {
		Prompt string `json:"prompt" binding:"required"`
	}

	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(400, gin.H{"message": "Prompt is required"})
		return
	}

	if len(req.Prompt) > 500 {
		c.JSON(400, gin.H{"message": "Prompt must be less than 500 characters"})
		return
	}

	usersCollection := database.GetDB().Collection("users")
	var user models.User
	err = usersCollection.FindOne(context.Background(), bson.M{"_id": userObjID}).Decode(&user)
	if err != nil {
		c.JSON(404, gin.H{"message": "User not found"})
		return
	}

	now := time.Now()
	lastReset := user.LogoGenerationCount.LastResetDate
	if now.Day() != lastReset.Day() || now.Month() != lastReset.Month() || now.Year() != lastReset.Year() {
		user.LogoGenerationCount.Count = 0
		user.LogoGenerationCount.LastResetDate = now
	}

	if user.LogoGenerationCount.Count >= h.DailyLimit {
		tomorrow := time.Date(now.Year(), now.Month(), now.Day()+1, 0, 0, 0, 0, now.Location())
		c.JSON(429, gin.H{
			"success":      false,
			"error":        "Daily limit reached",
			"message":      "You have reached your daily limit of logo generations",
			"limit":        h.DailyLimit,
			"used":         user.LogoGenerationCount.Count,
			"resetTime":    tomorrow.Format(time.RFC3339),
			"resetInHours": int(tomorrow.Sub(now).Hours()),
		})
		return
	}

	logoID := generateUUID()
	baseFilename := fmt.Sprintf("%s-%d-%s", userID, time.Now().Unix(), logoID)
	logosDir := filepath.Join(".", "uploads", "logos")
	os.MkdirAll(logosDir, 0755)
	filename := baseFilename + ".png"
	outputPath := filepath.Join(logosDir, filename)
	generationSource := "pollinations"

	pollinationsKey := strings.TrimSpace(os.Getenv("POLLINATIONS_API_KEY"))
	if pollinationsKey != "" {
		err = generateWithPollinations(req.Prompt, outputPath, pollinationsKey)
		if err != nil {
			log.Printf("Pollinations failed, falling back to SVG: %v", err)
			filename = baseFilename + ".svg"
			outputPath = filepath.Join(logosDir, filename)
			generationSource = "svg-fallback"
			err = generateSvgLogo(req.Prompt, outputPath)
		}
	} else {
		filename = baseFilename + ".svg"
		outputPath = filepath.Join(logosDir, filename)
		generationSource = "svg-fallback"
		err = generateSvgLogo(req.Prompt, outputPath)
	}
	if err != nil {
		c.JSON(500, gin.H{
			"success": false,
			"error":   "Generation failed",
			"message": err.Error(),
		})
		return
	}

	urlPath := "/uploads/logos/" + filename

	logoEntry := models.GeneratedLogo{
		LogoID:    logoID,
		URL:       urlPath,
		Prompt:    req.Prompt,
		FilePath:  filepath.Join("uploads", "logos", filename),
		CreatedAt: now,
		ExpiresAt: now.Add(7 * 24 * time.Hour),
	}

	user.LogoGenerationCount.Count++
	if len(user.GeneratedLogos) >= 20 {
		user.GeneratedLogos = user.GeneratedLogos[1:]
	}
	user.GeneratedLogos = append(user.GeneratedLogos, logoEntry)

	usersCollection.UpdateOne(context.Background(), bson.M{"_id": userObjID}, bson.M{
		"$set": bson.M{
			"generatedLogos":      user.GeneratedLogos,
			"logoGenerationCount": user.LogoGenerationCount,
		},
	})

	c.JSON(200, gin.H{
		"success":              true,
		"logo":                 logoEntry,
		"remainingGenerations": h.DailyLimit - user.LogoGenerationCount.Count,
		"source":               generationSource,
		"message":              "Logo generated successfully",
	})
}

func (h *LogoHandler) GetLogoHistory(c *gin.Context) {
	userID := c.GetString("userID")
	userObjID, err := primitive.ObjectIDFromHex(userID)
	if err != nil {
		c.JSON(400, gin.H{"message": "Invalid user ID"})
		return
	}

	usersCollection := database.GetDB().Collection("users")
	var user models.User
	err = usersCollection.FindOne(context.Background(), bson.M{"_id": userObjID}).Decode(&user)
	if err != nil {
		c.JSON(404, gin.H{"message": "User not found"})
		return
	}

	now := time.Now()
	var validLogos []models.GeneratedLogo
	for _, logo := range user.GeneratedLogos {
		if logo.ExpiresAt.After(now) {
			validLogos = append(validLogos, logo)
		}
	}

	for i, j := 0, len(validLogos)-1; i < j; i, j = i+1, j-1 {
		validLogos[i], validLogos[j] = validLogos[j], validLogos[i]
	}

	c.JSON(200, gin.H{
		"success":        true,
		"logos":          validLogos,
		"businessLogo":   user.BusinessLogo,
		"hasCustomLogo":  user.HasCustomLogo,
		"totalGenerated": len(user.GeneratedLogos),
	})
}

func (h *LogoHandler) GetLogoStatus(c *gin.Context) {
	userID := c.GetString("userID")
	userObjID, err := primitive.ObjectIDFromHex(userID)
	if err != nil {
		c.JSON(400, gin.H{"message": "Invalid user ID"})
		return
	}

	usersCollection := database.GetDB().Collection("users")
	var user models.User
	err = usersCollection.FindOne(context.Background(), bson.M{"_id": userObjID}).Decode(&user)
	if err != nil {
		c.JSON(404, gin.H{"message": "User not found"})
		return
	}

	now := time.Now()
	lastReset := user.LogoGenerationCount.LastResetDate
	if now.Day() != lastReset.Day() || now.Month() != lastReset.Month() || now.Year() != lastReset.Year() {
		user.LogoGenerationCount.Count = 0
	}

	tomorrow := time.Date(now.Year(), now.Month(), now.Day()+1, 0, 0, 0, 0, now.Location())

	c.JSON(200, gin.H{
		"success": true,
		"status": gin.H{
			"limit":        h.DailyLimit,
			"used":         user.LogoGenerationCount.Count,
			"remaining":    h.DailyLimit - user.LogoGenerationCount.Count,
			"resetTime":    tomorrow.Format(time.RFC3339),
			"resetInHours": int(tomorrow.Sub(now).Hours()),
		},
	})
}

func (h *LogoHandler) SelectLogo(c *gin.Context) {
	userID := c.GetString("userID")
	logoID := c.Param("logoId")

	userObjID, err := primitive.ObjectIDFromHex(userID)
	if err != nil {
		c.JSON(400, gin.H{"message": "Invalid user ID"})
		return
	}

	usersCollection := database.GetDB().Collection("users")
	var user models.User
	err = usersCollection.FindOne(context.Background(), bson.M{"_id": userObjID}).Decode(&user)
	if err != nil {
		c.JSON(404, gin.H{"message": "User not found"})
		return
	}

	var selectedLogo *models.GeneratedLogo
	for i, logo := range user.GeneratedLogos {
		if logo.LogoID == logoID || logo.ID.Hex() == logoID {
			if logo.ExpiresAt.After(time.Now()) {
				selectedLogo = &user.GeneratedLogos[i]
			} else {
				c.JSON(400, gin.H{
					"success": false,
					"error":   "Logo expired",
					"message": "This logo has expired. Please generate a new one.",
				})
				return
			}
			break
		}
	}

	if selectedLogo == nil {
		c.JSON(404, gin.H{
			"success": false,
			"error":   "Logo not found",
			"message": "The requested logo was not found in your history",
		})
		return
	}

	_, err = usersCollection.UpdateOne(context.Background(), bson.M{"_id": userObjID}, bson.M{
		"$set": bson.M{"businessLogo": selectedLogo.URL},
	})
	if err != nil {
		c.JSON(500, gin.H{"message": "Failed to select logo"})
		return
	}

	c.JSON(200, gin.H{
		"success":      true,
		"message":      "Logo selected as business logo",
		"businessLogo": selectedLogo.URL,
	})
}

func (h *LogoHandler) DeleteLogo(c *gin.Context) {
	userID := c.GetString("userID")
	logoID := c.Param("logoId")

	userObjID, err := primitive.ObjectIDFromHex(userID)
	if err != nil {
		c.JSON(400, gin.H{"message": "Invalid user ID"})
		return
	}

	usersCollection := database.GetDB().Collection("users")
	var user models.User
	err = usersCollection.FindOne(context.Background(), bson.M{"_id": userObjID}).Decode(&user)
	if err != nil {
		c.JSON(404, gin.H{"message": "User not found"})
		return
	}

	logoIndex := -1
	for i, logo := range user.GeneratedLogos {
		if logo.LogoID == logoID || logo.ID.Hex() == logoID {
			logoIndex = i
			break
		}
	}

	if logoIndex == -1 {
		c.JSON(404, gin.H{"message": "Logo not found"})
		return
	}

	logo := user.GeneratedLogos[logoIndex]
	if user.BusinessLogo != nil && *user.BusinessLogo == logo.URL {
		usersCollection.UpdateOne(context.Background(), bson.M{"_id": userObjID}, bson.M{
			"$set": bson.M{"businessLogo": nil},
		})
	}

	user.GeneratedLogos = append(user.GeneratedLogos[:logoIndex], user.GeneratedLogos[logoIndex+1:]...)
	usersCollection.UpdateOne(context.Background(), bson.M{"_id": userObjID}, bson.M{
		"$set": bson.M{"generatedLogos": user.GeneratedLogos},
	})

	os.Remove(logo.FilePath)

	c.JSON(200, gin.H{
		"success": true,
		"message": "Logo deleted successfully",
	})
}

func (h *LogoHandler) UploadCustomLogo(c *gin.Context) {
	userID := c.GetString("userID")
	userObjID, err := primitive.ObjectIDFromHex(userID)
	if err != nil {
		c.JSON(400, gin.H{"message": "Invalid user ID"})
		return
	}

	file, err := c.FormFile("logo")
	if err != nil {
		c.JSON(400, gin.H{"message": "No file uploaded"})
		return
	}

	if file.Size > 2*1024*1024 {
		c.JSON(400, gin.H{"message": "Logo must be less than 2MB"})
		return
	}

	logosDir := filepath.Join(".", "uploads", "logos")
	os.MkdirAll(logosDir, 0755)

	filename := fmt.Sprintf("%s-custom-%d%s", userID, time.Now().Unix(), filepath.Ext(file.Filename))
	filePath := filepath.Join(logosDir, filename)

	err = c.SaveUploadedFile(file, filePath)
	if err != nil {
		c.JSON(500, gin.H{"message": "Failed to save file"})
		return
	}

	usersCollection := database.GetDB().Collection("users")
	_, err = usersCollection.UpdateOne(context.Background(), bson.M{"_id": userObjID}, bson.M{
		"$set": bson.M{
			"businessLogo":  "/uploads/logos/" + filename,
			"hasCustomLogo": true,
		},
	})
	if err != nil {
		c.JSON(500, gin.H{"message": "Failed to update user"})
		return
	}

	c.JSON(200, gin.H{
		"success":      true,
		"message":      "Logo uploaded successfully",
		"businessLogo": "/uploads/logos/" + filename,
	})
}

func generateUUID() string {
	b := make([]byte, 16)
	rand.Read(b)
	return hex.EncodeToString(b)
}

func generateWithPollinations(prompt, outputPath, apiKey string) error {
	enhancedPrompt := fmt.Sprintf(
		"Professional minimalist logo design: %s, vector style, clean, modern, white background, suitable for business branding, high quality, no text",
		prompt,
	)
	seed := time.Now().UnixNano() % 1000000
	imageURL := fmt.Sprintf(
		"https://gen.pollinations.ai/image/%s?seed=%d&width=1024&height=1024&nologo=true&model=flux&negative_prompt=text,words,letters,watermark,signature,typography&key=%s",
		url.PathEscape(enhancedPrompt),
		seed,
		url.QueryEscape(apiKey),
	)

	client := &http.Client{Timeout: 120 * time.Second}
	req, err := http.NewRequest(http.MethodGet, imageURL, nil)
	if err != nil {
		return err
	}
	req.Header.Set("Accept", "image/*,*/*")
	req.Header.Set("User-Agent", "Mozilla/5.0 (compatible; MSMELogoBot/1.0)")

	resp, err := client.Do(req)
	if err != nil {
		return err
	}
	defer resp.Body.Close()

	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		return fmt.Errorf("pollinations request failed with status %d", resp.StatusCode)
	}

	contentType := strings.ToLower(resp.Header.Get("Content-Type"))
	if strings.HasPrefix(contentType, "text/") || strings.Contains(contentType, "json") {
		bodyPreview, _ := io.ReadAll(io.LimitReader(resp.Body, 512))
		return fmt.Errorf("pollinations returned non-image response (%s): %s", contentType, string(bodyPreview))
	}

	file, err := os.Create(outputPath)
	if err != nil {
		return err
	}
	defer file.Close()

	if _, err := io.Copy(file, resp.Body); err != nil {
		_ = os.Remove(outputPath)
		return err
	}

	return nil
}

func generateSvgLogo(prompt, outputPath string) error {
	colors := []struct{ bg, text, accent string }{
		{"#3B82F6", "#FFFFFF", "#1D4ED8"},
		{"#EF4444", "#FFFFFF", "#B91C1C"},
		{"#10B981", "#FFFFFF", "#047857"},
		{"#F59E0B", "#FFFFFF", "#D97706"},
		{"#8B5CF6", "#FFFFFF", "#6D28D9"},
	}

	palette := colors[0]

	stopWords := map[string]bool{
		"logo": true, "design": true, "for": true, "the": true, "and": true, "with": true,
		"a": true, "an": true, "of": true, "business": true,
	}
	parts := strings.Fields(strings.ToLower(prompt))
	words := make([]string, 0, len(parts))
	for _, part := range parts {
		clean := strings.Trim(part, ".,:;!?-_()[]{}\"'")
		if len(clean) > 1 && !stopWords[clean] {
			words = append(words, clean)
		}
	}

	initials := "LG"
	if len(words) >= 2 {
		initials = strings.ToUpper(string(words[0][0]) + string(words[1][0]))
	} else if len(words) == 1 {
		if len(words[0]) >= 2 {
			initials = strings.ToUpper(words[0][:2])
		} else {
			initials = strings.ToUpper(words[0])
		}
	}

	svg := `<?xml version="1.0" encoding="UTF-8"?>
<svg width="512" height="512" xmlns="http://www.w3.org/2000/svg">
  <defs>
    <linearGradient id="bgGrad" x1="0%" y1="0%" x2="100%" y2="100%">
      <stop offset="0%" style="stop-color:` + palette.bg + `;stop-opacity:1" />
      <stop offset="100%" style="stop-color:` + palette.accent + `;stop-opacity:1" />
    </linearGradient>
  </defs>
  <rect width="512" height="512" fill="url(#bgGrad)" rx="64" ry="64"/>
  <circle cx="256" cy="200" r="80" fill="` + palette.text + `"/>
  <text x="256" y="225" font-family="Arial, sans-serif" font-size="72" font-weight="bold" fill="` + palette.bg + `" text-anchor="middle" dominant-baseline="middle">` + initials + `</text>
  <text x="256" y="400" font-family="Arial, sans-serif" font-size="24" fill="` + palette.text + `" text-anchor="middle" fill-opacity="0.9">Business</text>
</svg>`

	f, err := os.Create(outputPath)
	if err != nil {
		return err
	}
	defer f.Close()

	_, err = f.WriteString(svg)
	return err
}
